### ISSUE TITLE
Custom plugin in Typescript after Vue 3.4 update

### ISSUE THREAD

**Issue by joaoefornazari** [vuejs/docs#2729]
[According to Vue 3.4 docs](https://vuejs.org/guide/reusability/plugins), if I want to define a custom plugin, I need to create it like this:

```[typescript]
// my-custom-plugin.ts
export default {
  install: (app, options) => {
    // do something here with app and options
  }
}
```

Then, on `main.ts`, I set app to use it:

```[typescript]
import App from './App.vue'
import router from './router'
import customPlugin from './my-custom-plugin'

export const app = createApp(App)
app.use(router)
app.use(customPlugin)
```

And that was OK until I updated Vue to version 3.4.20. I was using 3.2.45 before.

Now, when I try to build the project with `yarn build` (which uses `vite build` too) it throws this error:

```
Argument of type '{ install: (app: App, options: any) => void; }' is not assignable to parameter of type 'Plugin<[]>'.
  Type '{ install: (app: App, options: any) => void; }' is not assignable to type 'FunctionPlugin<[]>'.
```

After searching for type definitiions in `node_modules/@vue/runtime-core/dist/runtime-core.d.ts`, I found out that, in order to Typescript doesn't claim any type errors, I must define my plugin as an ObjectPlugin type.

I did it this way:

```
import { ObjectPlugin } from 'node_modules/@vue/runtime-core/dist/runtime-core'

export default {
  install: (app, options) => {
    // stuff
  },
} as ObjectPlugin

```

I can't use Plugin type because it says that `Plugin type is deprecated`.

Is my fix correct? How can I contribute so the Vue docs has some notes regarding Typescript support?

**Comment by NataliaTepluhina** [comment]
@joaoefornazari have you reached the issue on the core repository? It looks like we might improve the code instead of adding a specific section to docs 🤔

**Comment by joaoefornazari** [comment]
@NataliaTepluhina I'll give a try. This is a thing like that never happened to me so I was kind of lost in what to do about it - I thought that there was already something on Vue core that would solve this but people forgot to put it on the docs.

Maybe I open an issue on vuejs/core? What would you recommend?

**Comment by brc-dd** [comment]
> I can't use Plugin type because it says that `Plugin type is deprecated`.

Where? There is no deprecation notice for that. `import type { Plugin } from 'vue'` should work fine.

**Comment by joaoefornazari** [comment]
I solved the problem that happened with me here.

I did it like this:

- ` my-custom-plugin.ts`
```
// using import { PluginOptions } from 'unocss' also work
import { App, Plugin } from 'vue'

export default {
  install: (app: App, , options?: any) => {
    // stuff using app and options (optional)
  },
} satisfies Plugin
```

- `main.ts`
```
import App from './App.vue'
import router from './router'
import customPlugin from './my-custom-plugin'

export const app = createApp(App)
app.use(router)
app.use(customPlugin)
```

Thanks @brc-dd for the clarification!
But still, this solution ain't listed on the docs. @NataliaTepluhina Shouldn't it be there?

**Comment by os-tohe** [comment]
I have the exact same problem and the above solution did not work for me. I am using Vue version 3.4.26. I am creating a component library. The plugin works fine when I use it directly in my component library, but when I export the plugin and try to use it in another project it throws these errors:
```
Argument of type '{ install: (app: App<any>) => void; }' is not assignable to parameter of type 'Plugin<[]>'.
  Type '{ install: (app: App<any>) => void; }' is not assignable to type 'FunctionPlugin<[]>'.
    Type '{ install: (app: App<any>) => void; }' is not assignable to type '(app: App<any>) => any'.
      Type '{ install: (app: App<any>) => void; }' provides no match for the signature '(app: App<any>): any'.ts(2345)
```

**Comment by chriscdn** [comment]
I had this issue with v3.4.16, but upgrading to v3.4.29 resolved it.

**Comment by dmoczisko** [comment]
I'm also seeing this error when bumping to `vue version ^3.4.30`. Most recent working version is 3.4.25 - Variations of the solutions above do not resolve the problem.

**Comment by brc-dd** [comment]
Do any of you have a minimal, reproducible example?

**Comment by joaoefornazari** [comment]
@brc-dd Oh, sorry. I didn't see the comma lost over there. Also, the comma is not on my original code, so no problem at all - but I'll fix the code on the comment.

**Comment by brc-dd** [comment]
Yeah regarding docs I agree. But the last three comments are probably because of something else. They better be posted on discussions/discord. One is saying it works with v3.4.29 and the other is saying that most recent working version is v3.4.25 -- which are obviously conflicting. If any of those can share a repo where this happens we could probably resolve those.

**Comment by dmoczisko** [comment]
In my particular case, this issue was as a result of a mismatch in package.json files referencing vue

**Comment by luckylark2000** [comment]
I meet the same problem with v3.4.29 ,but upgrading to v3.4.30 resolved it。by the way, before I upgrading to v3.4.30, I delete the  node_modules and pnpm-lock.yaml, then I  reinstall the project by  `pnpm i` , maybe this also help 😄
